### PR TITLE
Add MD AGI subtraction for taxable social security

### LINKED ISSUE #1569: Missing MD AGI subtraction for social security benefits
The MD income tax has several AGI subtractions. One that is missing in the current code is the subtraction of federally taxable social security benefits. Note that on the [2021 Form 502R](https://www.marylandtaxes.gov/forms/21_forms/502R.pdf) this subtraction is calculated at the person level and the person-level subtraction amounts are then used in the calculation of the two-income subtraction.

From the 2021 Form 502 instructions we have this:

Screenshot 2022-12-16 at 7 08 46 AM

### LINKED ISSUE #1653: The taxable_social_security variable formula needs refinement
In the course of developing pull request #1652, a simple test failed unexpectedly.   Here is the test and its failure message:
```
- name: Test for MD social security AGI subtraction variable formula
  absolute_error_margin: 0.01
  period: 2021
  input:
    people:
      person1:
        is_tax_unit_head: true
        age: 62
        taxable_pension_income: 50_000
        social_security: 10_000
      person2:
        is_tax_unit_spouse: true
        age: 62
        social_security: 10_000
    spm_units:
      spm_unit:
        members: [person1, person2]
    tax_units:
      tax_unit:
        members: [person1, person2]
    households:
      household:
        members: [person1, person2]
        state_code: MD
  output:
    tax_unit_taxable_social_security: 17_000
    md_pension_subtraction_amount: 0
    md_socsec_subtraction: 17_000
    md_socsec_subtraction_amount: [8_500, 8_500]
```
The first three output variables passed, but the fourth failed unexpectedly.  Here is the test failure message:
```
Test 'Test for MD social security AGI subtraction variable formula':
md_socsec_subtraction_amount@2021: [17000.     0.] differs from [ 8500.  8500.]
  with an absolute margin [8500. 8500.] > 0.01
```
After reviewing the person-level `taxable_social_security` variable formula, I now understand why the test failed and why the `taxable_social_security` formula needs refinement.  Here are the details.

The `taxable_social_security` formula was developed eight months ago when the project was beginning to calculate US income taxes.  Here is the key comment from the formula's code:
```
def formula(person, period, parameters):
    # The taxable amount of Social Security is decided at the tax unit level, but
    # gross income (which contains taxable SS) is person-level. Therefore, we
    # include the taxable SS in gross income by assigning it to the head of the 
    # tax unit: this will not affect overall tax liability.
```
When that comment was written eight months ago, it was true, and is still true, if "overall tax liability" is understood as **US income tax liability**.   And that statement is true for some state income tax laws (for example, MO) that use the tax-unit sum of `taxable_social_security` rather than the person-level `taxable_social_security` amounts.

But for states that use the person-level `taxable_social_security` amounts to compute other person-level amounts (such as MD), the way in which the variable amount for `tax_unit_taxable_social_security` is allocated to people in the tax unit will "affect overall [state income] tax liability."

Given this situation, it seems to me that allocating the tax-unit amount to people in proportion to their total `social_security` amount would be a useful refinement.   Making this change would not affect US income tax liability and would not affect MO income tax liability.  But making this change would produce the correct MD income tax liability when a tax unit is eligible for the [two-income married couple AGI subtraction](https://www.marylandtaxes.gov/forms/21_forms/Resident_Booklet.pdf#page=16).  If we don't make this change, we will not be able to calculation MD AGI correctly in all cases.

### PR DESCRIPTION
**Pull request by martinholmer**
Fixes #1569.
Fixes #1653 by refining the person-level `taxable_social_security` variable formula.

### CODE REVIEW

File: policyengine_us/variables/gov/states/md/tax/income/agi/subtractions/md_socsec_subtraction_amount.py [hunk 1]
```diff
@@ -12,4 +12,32 @@ class md_socsec_subtraction_amount(Variable):
...
+        print("\n******************************************")
+        print("unit_tss=", unit_tss)
+        print("unit_socsec", unit_socsec)
+        print("ind_socsec", ind_socsec)
+        print("spouse_frac=", spouse_frac)
+        print("unit_spouse_frac=", unit_spouse_frac)
+        print("head_frac=", head_frac)
+        print("unit_head_frac=", unit_head_frac)
+        """
+        return select(
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/md/tax/income/agi/subtractions/md_socsec_subtraction_amount.py:40]: See https://github.com/PolicyEngine/policyengine-us/issues/1176
```suggestion
        return unit_tss * select([is_head, is_spouse], [unit_head_frac, unit_spouse_frac], default=0)
```

Comment (martinholmer) [reply policyengine_us/variables/gov/states/md/tax/income/agi/subtractions/md_socsec_subtraction_amount.py]: Done.  Thanks for the suggestion.

File: policyengine_us/variables/gov/states/md/tax/income/agi/subtractions/md_two_income_subtraction.py [hunk 2]
```diff
@@ -37,22 +37,33 @@ def formula(tax_unit, period, parameters):
...
-        subs_except_twoinc = [
-            sub for sub in p.sources if sub != "md_two_income_subtraction"
-        ]
-        total_subs_except_twoinc = add(tax_unit, period, subs_except_twoinc)
-        head_subs = 0.5 * total_subs_except_twoinc
-        spouse_subs = 0.5 * total_subs_except_twoinc
+        for sub in p.sources:
+            if sub == "md_two_income_subtraction":
+                continue
+            if sub == "md_pension_subtraction":
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/md/tax/income/agi/subtractions/md_two_income_subtraction.py:47]:
```suggestion
            # Person-level subtractions.
            if sub in ["md_pension_subtraction", "md_socsec_subtraction"]:
                ind_sub = person(sub + "_amount", period)
```

**Review by MaxGhenis — COMMENTED** [review]
Looks good just want to talk through the taxsim test removal (relating to the taxsim emulation branch) tomorrow before merging

File: policyengine_us/variables/gov/states/md/tax/income/agi/subtractions/md_socsec_subtraction.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    entity = TaxUnit
+    label = "MD socsec subtraction from AGI"
+    unit = USD
+    definition_period = YEAR
+    reference = (
+        "https://casetext.com/statute/code-of-maryland/article-tax-general/title-10-income-tax/subtitle-2-maryland-taxable-income-calculations-for-individual/part-ii-maryland-adjusted-gross-income/section-10-208-effective-until-712024-subtractions-from-federal-adjusted-gross-income-state-adjustments"
+        "https://www.marylandtaxes.gov/forms/21_forms/Resident_Booklet.pdf#page=14"
+    )
+    defined_for = StateCode.MD
+
```
[the remark this answers is not in the document]
Comment (martinholmer) [reply policyengine_us/variables/gov/states/md/tax/income/agi/subtractions/md_socsec_subtraction.py:15]: Done

File: policyengine_us/variables/gov/states/md/tax/income/agi/subtractions/md_pension_subtraction.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    entity = TaxUnit
+    label = "MD pension subtraction from AGI"
+    unit = USD
+    definition_period = YEAR
+    reference = (
+        "https://casetext.com/statute/code-of-maryland/article-tax-general/title-10-income-tax/subtitle-2-maryland-taxable-income-calculations-for-individual/part-ii-maryland-adjusted-gross-income/section-10-208-effective-until-712024-subtractions-from-federal-adjusted-gross-income-state-adjustments"
+        "https://www.marylandtaxes.gov/forms/21_forms/Resident_Booklet.pdf#page=13"
+    )
+    defined_for = StateCode.MD
+
```
[the remark this answers is not in the document]
Comment (martinholmer) [reply policyengine_us/variables/gov/states/md/tax/income/agi/subtractions/md_pension_subtraction.py:15]: Done

File: policyengine_us/parameters/gov/states/md/tax/income/agi/subtractions/pension_subtraction_max.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+Rdescription: MD maximum pension AGI subtraction
```
[the remark this answers is not in the document]
Comment (martinholmer) [reply policyengine_us/parameters/gov/states/md/tax/income/agi/subtractions/pension_subtraction_max.yaml:1]: Done.

File: policyengine_us/variables/gov/states/md/tax/income/agi/subtractions/md_socsec_subtraction.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+from policyengine_us.model_api import *
+
+
+class md_socsec_subtraction(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "MD socsec subtraction from AGI"
```
[the remark this answers is not in the document]
Comment (martinholmer) [reply policyengine_us/variables/gov/states/md/tax/income/agi/subtractions/md_socsec_subtraction.py:7]: Fixed.

File: policyengine_us/variables/gov/states/md/tax/income/agi/subtractions/md_two_income_subtraction.py (hunk 2 above)
Comment (martinholmer) [reply policyengine_us/variables/gov/states/md/tax/income/agi/subtractions/md_two_income_subtraction.py:47]: Changes made.

**Pull request merged**